dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,344 @@
+void NumberBox::ValidateInput()
+{
+    std::string InputAsString = winrt::to_string(m_TextBox.Text());
+    // Handles Empty TextBox Case, current behavior is to set Value to default (0)
+    if (InputAsString == "")
+    {
+        Value(0);
+        m_hasError = false;
+        return;
+    }
+
+    winrt::IReference<double> parsedNum = m_formatter.ParseDouble(m_TextBox.Text());
+    // TODO: Formatter.Format(value)
+
+    if (parsedNum && IsInBounds(parsedNum.Value()) )
+    {
+        SetErrorState(false);
+        Value(parsedNum.Value());
+        UpdateTextToValue();
+    }
+    else
+    {
+        SetErrorState(true);
+    }
+}
+
+// SpinClicks call to decrement or increment, 
+void NumberBox::OnSpinDownClick(winrt::IInspectable const&  sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(false);
+}
+void NumberBox::OnSpinUpClick(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    StepValue(true);
+}
+void NumberBox::OnNumberBoxKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args)
+{
+    switch (args.Key()) {
+        case winrt::VirtualKey::Up:
+        case winrt::VirtualKey::GamepadDPadUp:
+            StepValue(true);
+            break;
+        case winrt::VirtualKey::Down:
+        case winrt::VirtualKey::GamepadDPadDown:
+            StepValue(false);
+            break;
+    }
+}
+void NumberBox::OnScroll(winrt::IInspectable const& sender, winrt::PointerRoutedEventArgs const& args) {
+    if (!HyperScrollEnabled()) {
+        return;
+    }
+    int delta = args.GetCurrentPoint(*this).Properties().MouseWheelDelta();
+
+    if (delta > 0)
+    {
+        StepValue(true);
+    }
+    else if (delta < 0)
+    {
+        StepValue(false);
+    }
+}
+
+// Increments or decrements value by StepFrequency, wrapping if necessary
+void NumberBox::StepValue(bool sign)
+{
+    // Validating input before and after
+    ValidateInput();
+    double newVal = Value();
+
+    if (sign)
+    {
+        newVal = newVal + StepFrequency();
+    }
+    else
+    {
+        newVal =  newVal - StepFrequency();
+    }
+
+    // MinMaxMode Wrapping
+    if (MinMaxMode() == winrt::NumberBoxMinMaxMode::WrapEnabled && !IsInBounds(newVal))
+    {
+        while ( newVal > MaxValue() )
+        {
+            newVal = MinValue() + (newVal - MaxValue()) - 1;
+        }
+        while ( newVal < MinValue() )
+        {
+            newVal = MaxValue() - abs(newVal - MinValue()) + 1;
+        }
+        Value(newVal);
+        UpdateTextToValue();
+        return;
+    }
+
+    // Update Text and Revalidate new value
+    Value(newVal);
+    UpdateTextToValue();
+    ValidateInput();
+
+}
+
+// Runs formatter and updates TextBox to it's value property, run on construction if Value != 0
+void NumberBox::UpdateTextToValue()
+{
+        winrt::hstring formattedValue(m_formatter.Format(Value()));
+        Value( (m_formatter.ParseDouble(formattedValue)).Value() );
```

Comment:
Can you elaborate on this? I haven't seen any issues with using the formatter this way so far.

Reply:
User can enter a number with precision higher than shown - it's OK. It would be shown truncated, if NumberBox is configured so - it`s OK. But the original value should not be truncated till the user explicitly changes it.

Reply:
Thanks for the detailed explanation!